[Project] Add first task

Krekep opened the pull request:
- Add tests launch
- Add method for getting graph info
- Add method for create and export graph
- Add user console input handler
- Add tests for graphs method

Krekep committed:
[Project] Add first task

- Add tests launch
- Add method for getting graph info
- Add method for create and export graph
- Add user console input handler
- Add tests for graphs method

vadyushkins commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,51 @@
+      matrix:
+
+        # Each option you define in the matrix has a key and value
+        python-version: [ 3.8 ]
```
Добавьте в матрицу `ubuntu-latest`, `macos-latest`, `windows-latest`, чтобы тестировать ваше решение на всех платформах. Мы в задачах будем использовать только кросс-платформенные библиотеки, поэтому проблем не будет.

Krekep replied:
Исправлено.

vadyushkins commented on project/bridge.py:
```diff
@@ -0,0 +1,71 @@
+from pathlib import Path
+import networkx
```
```suggestion
import networkx as nx
```

Krekep replied:
А почему именно так? Просто потому что так принято?

Krekep replied:
Исправлено.

vadyushkins commented on project/__main__.py:
```diff
@@ -1 +1,5 @@
+
+if __name__ == "__main__":
+    print("exec sources directory")
+    console.run_console()
```
Запустил из корня репозитория `python3 -m project` и получил
```
AttributeError: module 'console' has no attribute 'run_console'
```
Кажется так не должно быть. Или я что-то сделал не так?

Krekep replied:
Исправлено.

vadyushkins commented on project/bridge.py:
```diff
@@ -0,0 +1,71 @@
+import pydot
+import utils
+
+__all__ = ["graph_info", "create_and_export", "shutdown"]
```
По заданию у вас должны быть следующие команды. (названия примерные)
```suggestion
command_names = ["get_graph_info", "create_two_cycles_graph", "save_to_dot", "quit"]
```

Krekep replied:
Исправлено.

vadyushkins commented on project/bridge.py:
```diff
@@ -0,0 +1,71 @@
+    Checks if the passed value is a file.
+    :param filename:
+    :return:
+    """
```
Используйте [`NumPy docstrings`](https://numpydoc.readthedocs.io/en/latest/format.html).
Минимально требуется: `1. Short summary`, `4. Parameters`, `5. Returns` или `6. Yields`

vadyushkins replied:
И так для всех функций и классов.

Krekep replied:
Исправлено.

vadyushkins commented on project/bridge.py:
```diff
@@ -0,0 +1,71 @@
+    print("Graph information:")
+    print("Number of nodes: ", info[0])
+    print("Number of edges: ", info[1])
+    print("Labels: ", *(info[2]))
```
Старайтесь разделять пыстуми строками осмысленные блоки текста, а не слепливать всё в одно полотно.

```suggestion
    if len(arg) != 1:
        raise Exception("Invalid amount of arguments!")
        
    filename = arg[0]
    name, extension = filename.split(".")
    
    if not _is_file(filename):
        raise Exception("No such file exists!")
        
    if extension != "dot":
        raise Exception("Wrong extension of file!")
        
    graph = _read_graph(filename)
    info = utils.get_graph_info(graph)
    
    print("Graph information:")
    print("Number of nodes: ", info[0])
    print("Number of edges: ", info[1])
    print("Labels: ", *(info[2]))
```

Krekep replied:
Исправлено.

vadyushkins commented on project/test.dot:
```diff
@@ -0,0 +1,35 @@
+digraph  {
```
Непонятно для чего используется этот файл. Да и папка `resources` в целом.
Если вы хотите написать тест, проверяющий, что вы правильно сохраняете файл в формате DOT, тот этот файл должен быть в папке `tests`. Желательно в `tests/data`.

Krekep replied:
Исправлено.

vadyushkins commented on requirements.txt:
```diff
@@ -1,6 +1,7 @@
 antlr4-python3-runtime
 black
 cfpq-data
+networkx
```
`networkx` уже есть в зависимостях у `cfpq_data`. Почему вы его добавили?
```suggestion
```

Krekep replied:
Исправлено.

vadyushkins commented on tests/test_utils.py:
```diff
@@ -0,0 +1,27 @@
+
+
+def teardown_module(module):
+    print("basic teardown module")
```
Кажется здесь этого быть не должно

Krekep replied:
Исправлено.

vadyushkins commented on tests/test_utils.py:
```diff
@@ -0,0 +1,27 @@
+    g_info = utils.get_graph_info(graph)
+    assert g_info[0] == 16
+    assert g_info[1] == 17
+    assert g_info[2] == {"a", "b"}
```
В unit тестах лучше придерживаться правила один assert на тест.

Krekep replied:
Исправлено.

vadyushkins commented on tests/test_utils.py:
```diff
@@ -0,0 +1,27 @@
+    assert g_info[2] == {"a", "b"}
+
+
+def test_saving_graph():
+    utils.generate_and_export_two_cycle(
+        10, 5, ("a", "b"), "tests/data/example_saving.dot"
+    )
```
Что этот тест проверяет?
Как минимум стоит проверять то, что тот граф, который вы сохраняете, совпадает с тем, что вы ожидаете увидеть как он сохранится. По сути должен быть `graph_expected.dot` в `tests/data`.

Krekep replied:
Исправлено.

vadyushkins reviewed (changes requested):
Также есть пара глобальных замечаний.

* Запускаться `REPL` должен через `python -m project`. Сейчас так не работает.

* >   - [ ] По имени графа вернуть количество вершин, рёбер и перечислить различные метки, встречающиеся на рёбрах.

    Именно **по имени**. В наборе данных у графов есть имена. Такие как `geospecies`, `pizza`, `enzyme`. См. [Get a real graph](https://jetbrains-research.github.io/CFPQ_Data/tutorial.html#get-a-real-graph)
    
* Копируйте из `tasks` полностью описание задачи в комментарий вашего PR и проставляйте чекбоксики, пожалуйста. Так будет и вам и мне проще видеть, что требовалось сделать и что уже сделано.

* Не забудьте написать в `project/README.md` как запускать ваше **консольное приложение**.

* Убедитесь, что ваши тесты покрывают все возможные крайние случаи.
    
> Мой комментарий на русском языке, так как курс читается на русском языке.
Если вы согласны перейти на английский, напишите мне об этом, пожалуйста, в комментариях.

Krekep committed:
[Fix] Fix the notes

- Fix function description
- Fix start application
- Fix tests
- Add create_two_cycles method
- Add save_to_dot method
- Remove create_and_export_two_cycles method
- Remove extra files

Krekep commented:
# Задача 1. Инициализация рабочего окружения

* **Мягкий дедлайн**: 05.09.2021, 23:59
* **Жёсткий дедлайн**: 08.09.2021, 23:59
* Полный балл: 5

## Задача

- [x] Сделать публичный `fork` данного репозитория.
- [x] Добавить ссылку на ваш `fork` в [таблицу](https://docs.google.com/spreadsheets/d/18DhYG5CuOrN4A5b5N7-mEDfDkc-7BuXF3Qsu6HD-lks/edit?usp=sharing).
- [x] Добавить в совладельцы форка одного из ассистентов (чтобы узнать, кого именно, нужно посмотреть в [таблицу](https://docs.google.com/spreadsheets/d/18DhYG5CuOrN4A5b5N7-mEDfDkc-7BuXF3Qsu6HD-lks/edit?usp=sharing))
- [x] Реализовать консольное приложение, предоставляющее перечисленные ниже возможности. Для работы с графами использовать [cfpq-data](https://jetbrains-research.github.io/CFPQ_Data/tutorial.html#graphs). Данное приложение в дальнейшем будет расширяться. Функции приложения, которые необходимо реализовать:
  - [x] По имени графа вернуть количество вершин, рёбер и перечислить различные метки, встречающиеся на рёбрах.
  - [x] По количеству вершин в циклах и именам меток строить [граф из двух циклов](https://jetbrains-research.github.io/CFPQ_Data/reference/graphs/generated/cfpq_data.graphs.generators.labeled_two_cycles_graph.html#cfpq_data.graphs.generators.labeled_two_cycles_graph) и сохранять его в указанный файл в формате DOT (использовать pydot).
- [x] Добавить необходимые тесты.
- [x] Добавить запуск тестов с помощью `pytest` в `.github/workflows/`.

Krekep committed:
[Readme] Add readme to project

vadyushkins reviewed (approved):
Молодец :+1:

> Советую, чтобы не сохранять все исправления в истории, сделать `Squash and merge`

Замечания на будущее:
1. Старайтесь в `docstrings` разделять блоки `Parameters`, `Returns`, etc.
2. Исправьте `from . import ...` на [правильный вариант по PEP8](https://www.python.org/dev/peps/pep-0008/#imports)
